packages/@aws-cdk/core/lib/stack.ts

```diff
@@ -1344,9 +1427,26 @@ function generateExportName(stackExports: Construct, id: string) {
   return prefix + localPart.slice(Math.max(0, localPart.length - maxLength + prefix.length));
 }
 
+
+function filterReasons(dependency: StackDependency, reasonFilter: StackDependencyReason) {
+  let reasons: Set<StackDependencyReason> = new Set();
+  dependency.reasons.forEach((reason) => {
+    if ((!reasonFilter.source || reasonFilter.source == reason.source) && (!reasonFilter.target || reasonFilter.target == reason.target)) {
+      reasons.add(reason);
+    }
+  });
+  return Array.from(reasons);
+}
+
+interface StackDependencyReason {
+  description : string;
+  source?: Element;
+  target?: Element;
+}
+
 interface StackDependency {
   stack: Stack;
-  reasons: string[];
+  reasons: StackDependencyReason[];
```

Comment:
I'm not sure I get what benefit you're getting from this change.

Reply:
So, in order add a more complete add/replace/remove dependency interface, we need a clearer way to automatically look up and potentially remove a given stack dependency reason. Consider a case where we have two stacks, 1, and 2, each with resources A, B and C, D respectively:
```yaml
Stack1:
  - ResourceA
  - ResourceB
Stack2
  - ResourceC
  - ResourceD
```

If user code does both `ResourceA.addDependsOn(ResourceC)` and `ResourceB.addDependsOn(RecourceD)`, then it will result in only one stack dependency: `Stack1` on `Stack2`, but with two independent instances of `StackDependencyReason`. We need an unambiguous way to track those reasons so that the new `removeDependsOn` can appropriately back those dependencies back out, if the user desires. So if user code later does `ResourceA.removeDependsOn(ResourceC)`, the method will resolve it to a stack-level dependency and, rather than remove the dependency, it will find the matching `source` and `target` in the `StackDependencyReason`, remove that reason, but leave the stack dependency in place, since it will still have the `source: ResourceB, target: ResourceD` dependency reason. If the user code also then does `ResourceB.removeDependsOn(ResourceD)`, the method will again resolve to the stack level, remove the corresponding `StackDependencyReason` then, seeing that there is no longer a reason for the dependency, remove the `Stack1` on `Stack2` dependency as well.

Does that make sense? Without a clear data structure for automated reason matching, appropriately tracking resource-to-resource dependencies across stacks, we really can't be intelligent about removal in the future.